File: sdk/dart/lib/src/abac.dart
/// ABAC operations for the GGID Dart SDK.
///
/// The core ABAC methods (evaluateAbac, checkPolicy) are already available
/// as instance methods on [GGIDClient].
///
/// This file provides extension methods for additional ABAC utilities.
library ggid.abac;

import 'package:ggid/ggid_client.dart';
import 'models.dart';

/// Extension providing convenience ABAC operations.
extension AbacExtension on GGIDClient {
  /// Convenience: check policy with simple parameters.
  ///
  /// ```dart
  /// final allowed = await ggid.checkPolicySimple(
  ///   token, 'user-1', 'documents', 'read',
  ///   context: {'department': 'finance'},
  /// );
  /// ```
  Future<bool> checkPolicySimple(
    String token,
    String subject,
    String resource,
    String action, {
    Map<String, String> context = const {},
  }) async {
    final result = await checkPolicy(token, PolicyCheckRequest(
      subject: subject,
      resource: resource,
      action: action,
      context: context,
    ));
    return result.allowed;
  }

  /// Evaluate ABAC with a simple condition list.
  Future<AbacEvalResult> evaluateAbacSimple(
    String token,
    String resource,
    String action,
    List<AbacCondition> conditions,
  ) async {
    return evaluateAbac(token, AbacEvalRequest(
      action: action,
      resource: resource,
      conditions: conditions,
    ));
  }
}

File: sdk/csharp/src/ABAC.cs
using GGID.SDK.Models;

namespace GGID.SDK;

/// <summary>
/// ABAC operations: attribute-based policy evaluation.
/// </summary>
public static class ABAC
{
    /// <summary>
    /// Evaluate an ABAC policy with conditions.
    /// </summary>
    public static async Task<AbacEvalResult> EvaluateAbacAsync(this GGIDClient client, string token, AbacEvalRequest request, CancellationToken ct = default)
    {
        return await client.PostAsync<AbacEvalResult>("/api/v1/policies/abac/evaluate", request, token, ct);
    }

    /// <summary>
    /// Check a policy with subject, resource, action, and context.
    /// </summary>
    public static async Task<PolicyResult> CheckPolicyAsync(this GGIDClient client, string token, PolicyCheckRequest request, CancellationToken ct = default)
    {
        return await client.PostAsync<PolicyResult>("/api/v1/policies/check", request, token, ct);
    }

    /// <summary>
    /// Convenience: check policy with simple parameters.
    /// </summary>
    public static async Task<bool> CheckPolicyAsync(this GGIDClient client, string token, string subject, string resource, string action, Dictionary<string, string>? context = null, CancellationToken ct = default)
    {
        var req = new PolicyCheckRequest
        {
            Subject = subject,
            Resource = resource,
            Action = action,
            Context = context ?? new(),
        };
        var result = await client.CheckPolicyAsync(token, req, ct);
        return result?.Allowed ?? false;
    }
}
